### ISSUE TITLE
[Trouble Shooting] Valid score가 가끔씩 심하게 요동치는 문제 요약

### ISSUE THREAD

**Issue by jkim1209** [AIBootcamp14/computervisioncompetition-cv2#31]
## 🐞 문제 상황

- 발생 시각: 지금까지 몇 번 Valid Loss 및 Valid F1 score가 출렁이는 현상이 발견됨 - 문제가 있는 것으로 이제야 파악됨 (Valid F1 score 가 1을 달성하는 문제 발생)
- 문제 상황 설명: 
 현재 참가자가 소유한 Train data는 깨끗한 데이터(이미지)이고, Test data는 상대적으로 의도적으로 훼손된 데이터(이미지) 입니다. 따라서 Train data로 모델을 학습시킬 때 Train과 Valid set으로 나누는데, Train data는 Test data와 비슷하게 만들기 위해 의도적으로 훼손시키는 방향으로 변형(transform)시킨 뒤 학습시킵니다. 
 문제는 Valid set입니다. 전통적인 방법에 따라 Valid set은 Test set과 같은 방식의 변형(transform)을 사용한 뒤 평가를 하면 깨끗한 Valid set과 훼손된 Test set과의 괴리가 커 Test set의 결과를 제대로 예측하지 못하는 문제가 발생합니다. 
 따라서 여태까지 Valid set에는 Train set보다는 훼손정도가 덜 한 정도의 '약한 훼손'을 주는 정도로 별도의 변형을 준 뒤(`build_valid_tf()`) 모델 학습을 진행했습니다. 
 하지만 이러한 변형은 epoch마다 적용됩니다. 즉, epoch마다 valid data는 random하게 바뀌어 같은 fold 내에서도 매 epoch마다 모델은 다른 데이터에 대해 평가하고 점수를 매기는 셈이 됩니다. 그러므로 이러한 valid set 을 기반으로 출력된 valid_loss 및 f1 score는 재현 불가능할 뿐만 아니라 일관성이 없어 적절하지 않습니다.

- 관련 코드/명령어:

```python
# src/data/transform.py
# ...

def build_valid_tf(size, mean, std, interpolation=cv2.INTER_LINEAR):
    return A.Compose([
        A.Rotate(limit=5, border_mode=0, value=255, p=0.3),
        A.ImageCompression(quality_lower=50, quality_upper=95, p=0.2),
        A.ToGray(p=0.3),
        *_finalize_to_square(size, interpolation),
        A.Normalize(mean=mean, std=std),
        ToTensorV2()
    ])

# src/train/train.py
# ...

def run_train( ...
    ...
    for fold, ...
        ....
        valid_tf_local = build_valid_tf(img_size, mean, std, interpolation=interp)
```

---

## 🔍 해결책

- [ ] 각 fold 마다 한 번만 `bild_valid_tf`를 실행하고, 이를 캐시 데이터로 저장합니다. 이후 같은 fold 내에서는 valid set을 불러올 때마다 캐시된 데이터를 불러와서 평가합니다. 이러한 방식으로 한 fold 내에서 모델은 매 epoch마다 '똑같은' valid set에 대해 평가합니다.
  - 코드에서는 `CachedDataset` 부분으로 구현

---

## ✅ 해결되었는지 로그 확인

- 아래는 위의 변경점을 반영한 뒤 하나의 예시입니다. 더 이상 valid_loss 및 f1 score가 출렁이지 않고 일정수준까지 안정적으로 감소(loss)/증가(f1)하는 패턴을 보임을 알 수 있습니다. 이후에는 전형적인 Overfitting pattern에 부합하게 Valid loss가 일관되게 증가하는 패턴을 보입니다.

```bash
========== Fold 1/5 ==========
[Fold 1] Epoch 01 | train_loss 2.7747 f1 0.1415 || valid_loss 2.7971 f1 0.1062
[Fold 1] Epoch 02 | train_loss 2.5470 f1 0.3200 || valid_loss 2.7906 f1 0.1208
[Fold 1] Epoch 03 | train_loss 2.1725 f1 0.5000 || valid_loss 2.7755 f1 0.1295
[Fold 1] Epoch 04 | train_loss 1.5900 f1 0.6540 || valid_loss 2.7501 f1 0.1492
[Fold 1] Epoch 05 | train_loss 1.0032 f1 0.8278 || valid_loss 2.7153 f1 0.1762
[Fold 1] Epoch 06 | train_loss 0.5733 f1 0.8642 || valid_loss 2.6730 f1 0.2085
[Fold 1] Epoch 07 | train_loss 0.3588 f1 0.9177 || valid_loss 2.6246 f1 0.2539
[Fold 1] Epoch 08 | train_loss 0.2472 f1 0.9336 || valid_loss 2.5704 f1 0.3029
[Fold 1] Epoch 09 | train_loss 0.2120 f1 0.9443 || valid_loss 2.5111 f1 0.3648
[Fold 1] Epoch 10 | train_loss 0.1635 f1 0.9690 || valid_loss 2.4470 f1 0.4282
[Fold 1] Epoch 11 | train_loss 0.1384 f1 0.9795 || valid_loss 2.3772 f1 0.4809
[Fold 1] Epoch 12 | train_loss 0.1249 f1 0.9809 || valid_loss 2.3017 f1 0.5123
[Fold 1] Epoch 13 | train_loss 0.1066 f1 0.9906 || valid_loss 2.2197 f1 0.5426
[Fold 1] Epoch 14 | train_loss 0.1090 f1 0.9923 || valid_loss 2.1323 f1 0.5958
[Fold 1] Epoch 15 | train_loss 0.1040 f1 0.9911 || valid_loss 2.0388 f1 0.6328
[Fold 1] Epoch 16 | train_loss 0.1068 f1 0.9920 || valid_loss 1.9403 f1 0.6916
[Fold 1] Epoch 17 | train_loss 0.1099 f1 0.9911 || valid_loss 1.8344 f1 0.7344
[Fold 1] Epoch 18 | train_loss 0.1012 f1 0.9919 || valid_loss 1.7251 f1 0.7582
[Fold 1] Epoch 19 | train_loss 0.1013 f1 0.9925 || valid_loss 1.6131 f1 0.7895
[Fold 1] Epoch 20 | train_loss 0.1071 f1 0.9907 || valid_loss 1.4983 f1 0.8164
[Fold 1] Epoch 21 | train_loss 0.1190 f1 0.9899 || valid_loss 1.3818 f1 0.8329
[Fold 1] Epoch 22 | train_loss 0.0933 f1 0.9963 || valid_loss 1.2662 f1 0.8457
[Fold 1] Epoch 23 | train_loss 0.0883 f1 0.9961 || valid_loss 1.1526 f1 0.8570
[Fold 1] Epoch 24 | train_loss 0.0883 f1 0.9971 || valid_loss 1.0438 f1 0.8798
[Fold 1] Epoch 25 | train_loss 0.0831 f1 0.9984 || valid_loss 0.9368 f1 0.8921
[Fold 1] Epoch 26 | train_loss 0.0927 f1 0.9957 || valid_loss 0.8391 f1 0.9081
[Fold 1] Epoch 27 | train_loss 0.0825 f1 0.9985 || valid_loss 0.7492 f1 0.9132
[Fold 1] Epoch 28 | train_loss 0.0844 f1 0.9984 || valid_loss 0.6663 f1 0.9204
[Fold 1] Epoch 29 | train_loss 0.0868 f1 0.9986 || valid_loss 0.5931 f1 0.9346
[Fold 1] Epoch 30 | train_loss 0.0931 f1 0.9965 || valid_loss 0.5282 f1 0.9310
[Fold 1] Epoch 31 | train_loss 0.0974 f1 0.9944 || valid_loss 0.4708 f1 0.9334
[Fold 1] Epoch 32 | train_loss 0.0827 f1 0.9984 || valid_loss 0.4216 f1 0.9383
[Fold 1] Epoch 33 | train_loss 0.0850 f1 0.9974 || valid_loss 0.3800 f1 0.9495
[Fold 1] Epoch 34 | train_loss 0.1025 f1 0.9933 || valid_loss 0.3441 f1 0.9541
[Fold 1] Epoch 35 | train_loss 0.0882 f1 0.9975 || valid_loss 0.3143 f1 0.9571
[Fold 1] Epoch 36 | train_loss 0.0847 f1 0.9982 || valid_loss 0.2880 f1 0.9607
[Fold 1] Epoch 37 | train_loss 0.0821 f1 0.9992 || valid_loss 0.2653 f1 0.9607
[Fold 1] Epoch 38 | train_loss 0.0917 f1 0.9968 || valid_loss 0.2462 f1 0.9576
[Fold 1] Epoch 39 | train_loss 0.0881 f1 0.9983 || valid_loss 0.2303 f1 0.9576
[Fold 1] Epoch 40 | train_loss 0.0851 f1 0.9984 || valid_loss 0.2167 f1 0.9570
[Fold 1] Epoch 41 | train_loss 0.0828 f1 0.9993 || valid_loss 0.2054 f1 0.9614
[Fold 1] Epoch 42 | train_loss 0.0896 f1 0.9976 || valid_loss 0.1955 f1 0.9614
[Fold 1] Epoch 43 | train_loss 0.0848 f1 0.9993 || valid_loss 0.1866 f1 0.9689
[Fold 1] Epoch 44 | train_loss 0.0980 f1 0.9952 || valid_loss 0.1795 f1 0.9689
[Fold 1] Epoch 45 | train_loss 0.0924 f1 0.9969 || valid_loss 0.1734 f1 0.9689
[Fold 1] Epoch 46 | train_loss 0.0815 f1 0.9992 || valid_loss 0.1652 f1 0.9689
[Fold 1] Epoch 47 | train_loss 0.0844 f1 0.9993 || valid_loss 0.1600 f1 0.9689
[Fold 1] Epoch 48 | train_loss 0.0854 f1 0.9981 || valid_loss 0.1560 f1 0.9751
[Fold 1] Epoch 49 | train_loss 0.0797 f1 1.0000 || valid_loss 0.1520 f1 0.9781
[Fold 1] Epoch 50 | train_loss 0.0870 f1 0.9984 || valid_loss 0.1494 f1 0.9781
[Fold 1] Epoch 51 | train_loss 0.0849 f1 0.9993 || valid_loss 0.1477 f1 0.9810
[Fold 1] Epoch 52 | train_loss 0.0797 f1 1.0000 || valid_loss 0.1451 f1 0.9810
[Fold 1] Epoch 53 | train_loss 0.0840 f1 0.9984 || valid_loss 0.1429 f1 0.9810
[Fold 1] Epoch 54 | train_loss 0.0884 f1 0.9983 || valid_loss 0.1407 f1 0.9844
[Fold 1] Epoch 55 | train_loss 0.0885 f1 0.9975 || valid_loss 0.1383 f1 0.9843
[Fold 1] Epoch 56 | train_loss 0.0888 f1 0.9975 || valid_loss 0.1384 f1 0.9878
[Fold 1] Epoch 57 | train_loss 0.0825 f1 0.9991 || valid_loss 0.1383 f1 0.9878
[Fold 1] Epoch 58 | train_loss 0.0924 f1 0.9971 || valid_loss 0.1372 f1 0.9878
[Fold 1] Epoch 59 | train_loss 0.0933 f1 0.9956 || valid_loss 0.1390 f1 0.9878
[Fold 1] Epoch 60 | train_loss 0.0809 f1 0.9984 || valid_loss 0.1405 f1 0.9878
[Fold 1] Epoch 61 | train_loss 0.0904 f1 0.9961 || valid_loss 0.1407 f1 0.9878
[Fold 1] Epoch 62 | train_loss 0.0867 f1 0.9966 || valid_loss 0.1423 f1 0.9878
[Fold 1] Epoch 63 | train_loss 0.0910 f1 0.9970 || valid_loss 0.1433 f1 0.9878
[Fold 1] Epoch 64 | train_loss 0.0839 f1 0.9983 || valid_loss 0.1444 f1 0.9878
[Fold 1] Epoch 65 | train_loss 0.0828 f1 0.9992 || valid_loss 0.1460 f1 0.9878
[Fold 1] Epoch 66 | train_loss 0.0803 f1 0.9992 || valid_loss 0.1478 f1 0.9878
[Fold 1] Epoch 67 | train_loss 0.0795 f1 1.0000 || valid_loss 0.1497 f1 0.9878
[Fold 1] Epoch 68 | train_loss 0.0892 f1 0.9985 || valid_loss 0.1511 f1 0.9878
[Fold 1] Epoch 69 | train_loss 0.0818 f1 0.9993 || valid_loss 0.1528 f1 0.9878
[Fold 1] Epoch 70 | train_loss 0.0853 f1 0.9984 || valid_loss 0.1538 f1 0.9878
[Fold 1] Epoch 71 | train_loss 0.0833 f1 0.9974 || valid_loss 0.1541 f1 0.9878
[Fold 1] Early stopping triggered at epoch 71. (best_valid_f1=0.9878)
Valid Fold1 mismatch count by class: {11: 1, 7: 1, 3: 1, 4: 1}
```

---
